host/xtest/pkcs11_1000.c

```diff
@@ -8744,6 +8735,126 @@ static void xtest_pkcs11_test_1027(ADBG_Case_t *c)
+		{ CKA_MODIFIABLE, &(CK_BBOOL){ CK_TRUE }, sizeof(CK_BBOOL) },
+		{ CKA_CLASS, &(CK_OBJECT_CLASS){ CKO_SECRET_KEY }, sizeof(CK_OBJECT_CLASS) },
+		{ CKA_KEY_TYPE, &(CK_KEY_TYPE){ CKK_AES }, sizeof(CK_KEY_TYPE) },
+	};
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_OBJECT_HANDLE obj_handle[2] = { };
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_ULONG hdl_count = 0;
+	CK_ULONG obj_size = 0;
+	CK_SLOT_ID slot = 0;
+	int ret = 1;
+
+	rv = init_lib_and_find_token_slot(&slot, PIN_AUTH);
+	if (rv != CKR_OK)
+		return 1;
+
+	rv = C_OpenSession(slot, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL, 0, &session);
+	if (rv != CKR_OK)
+		goto out_lib;
+
+	/*
+	 * We expect to find a single matching object so that it is
+	 * the very one we're about to destroy.
+	 */
+	rv = C_FindObjectsInit(session, find_token_template, ARRAY_SIZE(find_token_template));
+	if (rv != CKR_OK)
+		goto out_session;
+
+	rv = C_FindObjects(session, obj_handle, ARRAY_SIZE(obj_handle), &hdl_count);
+	if (rv != CKR_OK || hdl_count != 1)
+		goto out_session;
+
+	rv = C_FindObjectsFinal(session);
+	if (rv != CKR_OK)
+		goto out_session;
+
+	/* Session B destroys the token object, session A shall not reach it */
+	rv = C_DestroyObject(session, obj_handle[0]);
+	if (rv != CKR_OK)
+		goto out_session;
+
+	rv = C_GetObjectSize(session, obj_handle[0], &obj_size);
+	if (rv == CKR_OBJECT_HANDLE_INVALID)
+		ret = 0;
+
+out_session:
+	C_CloseSession(session);
+out_lib:
+	close_lib();
+
+	return ret;
+}
+
+/*
+ * This test invloves 2 client sessions towards the PKCS11 token.
+ *
+ * Session A (implementation below) creates a token object then executes an
+ * application that opens session B to destroy the object after what session A
+ * gets a invalid object handle result when reusing its object handle.
+ *
+ * Session B sequence is implemented by function
+ * xtest_pkcs11_1028_destroy_token_object() that must be called from another
+ * process so that cryptoki library sees it as a different client and
+ * generates another handle for the same object.
+ */
+static void xtest_pkcs11_test_1028(ADBG_Case_t *c)
+{
+	CK_SESSION_HANDLE session = CK_INVALID_HANDLE;
+	CK_OBJECT_HANDLE obj_handle = CK_INVALID_HANDLE;
+	CK_RV rv = CKR_GENERAL_ERROR;
+	CK_ULONG obj_size = 0;
+	CK_SLOT_ID slot = 0;
+	int ret = 0;
+
+	rv = init_lib_and_find_token_slot(&slot, PIN_AUTH);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		return;
+
+	/* Session A creates a token object */
+	rv = C_OpenSession(slot, CKF_SERIAL_SESSION | CKF_RW_SESSION, NULL, 0, &session);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out_lib;
+
+	rv = C_CreateObject(session, cktest_token_object, ARRAY_SIZE(cktest_token_object),
+			    &obj_handle);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out_session;
+
+	rv = C_GetObjectSize(session, obj_handle, &obj_size);
+	if (!ADBG_EXPECT_CK_OK(c, rv)) {
+		rv = C_DestroyObject(session, obj_handle);
+		ADBG_EXPECT_CK_OK(c, rv);
+		goto out_session;
+	}
+
+	ret = system("xtest --pkcs11-1028-destroy-token-object");
```

Comment:
I wonder if it wouldn't be better to use what's in `argv[0]` than assuming that `xtest` is OK.

Reply:
Nice catch. Maybe export `argv[0]` from _xtest_main.c_ with some global variable (`char *xtest_argv_0`)?

Reply:
Sure